File: archive/polyglot/go/weather_risk.go
// Package weatherrisk provides weather-aware risk scoring for SDACS.
package weatherrisk

import "math"

type Input struct {
	WindMps            float64
	VisibilityKm       float64
	PrecipitationLevel float64
	Congestion         float64
}

type Output struct {
	Score    float64
	Category string
}

func clamp(v, lo, hi float64) float64 {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func Compute(input Input) Output {
	windNorm := clamp(input.WindMps/20.0, 0.0, 1.0)
	visNorm := clamp(input.VisibilityKm/10.0, 0.0, 1.0)
	visPenalty := 1.0 - visNorm
	precip := clamp(input.PrecipitationLevel, 0.0, 1.0)
	congestion := clamp(input.Congestion, 0.0, 1.0)

	raw := 0.35*windNorm + 0.25*visPenalty + 0.20*precip + 0.20*congestion
	score := math.Round(clamp(raw, 0.0, 1.0)*10000.0) / 10000.0

	category := "RED"
	if score < 0.25 {
		category = "GREEN"
	} else if score < 0.50 {
		category = "YELLOW"
	} else if score < 0.75 {
		category = "ORANGE"
	}

	return Output{Score: score, Category: category}
}

File: archive/polyglot/rust/weather_risk.rs
//! SDACS weather risk model (Rust).
//! Keeps parity with Python canonical formula.

#[derive(Debug, Clone, Copy)]
pub struct WeatherRiskInput {
    pub wind_mps: f64,
    pub visibility_km: f64,
    pub precipitation_level: f64,
    pub congestion: f64,
}

#[derive(Debug, Clone)]
pub struct WeatherRiskOutput {
    pub score: f64,
    pub category: &'static str,
}

fn clamp(v: f64, lo: f64, hi: f64) -> f64 {
    if v < lo {
        lo
    } else if v > hi {
        hi
    } else {
        v
    }
}

pub fn compute_weather_risk(input: WeatherRiskInput) -> WeatherRiskOutput {
    let wind_norm = clamp(input.wind_mps / 20.0, 0.0, 1.0);
    let vis_norm = clamp(input.visibility_km / 10.0, 0.0, 1.0);
    let vis_penalty = 1.0 - vis_norm;
    let precip = clamp(input.precipitation_level, 0.0, 1.0);
    let congestion = clamp(input.congestion, 0.0, 1.0);

    let raw = 0.35 * wind_norm + 0.25 * vis_penalty + 0.20 * precip + 0.20 * congestion;
    let score = (clamp(raw, 0.0, 1.0) * 10000.0).round() / 10000.0;

    let category = if score < 0.25 {
        "GREEN"
    } else if score < 0.50 {
        "YELLOW"
    } else if score < 0.75 {
        "ORANGE"
    } else {
        "RED"
    };

    WeatherRiskOutput { score, category }
}
